mayastor/tests/rebuild.rs

```diff
@@ -0,0 +1,1051 @@
+            .await
+            .expect("Failed to resume rebuild");
+        assert!(
+            wait_for_rebuild_completion(
+                nexus_hdl,
+                &child.share_uri,
+                Duration::from_secs(10),
+            )
+            .await
+        );
+        remove_child(nexus_hdl, &child.share_uri).await;
+    }
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+
+    // Add the children back again
+    for child in &degraded_children {
+        add_child(nexus_hdl, &child.share_uri, true).await;
+    }
+
+    // Wait for rebuilds to complete
+    for child in &degraded_children {
+        assert!(
+            wait_for_rebuild_completion(
+                nexus_hdl,
+                &child.share_uri,
+                Duration::from_secs(10),
+            )
+            .await
+        );
+    }
+}
+
+/// Test rebuild while running front-end I/O.
+/// Note: This test can take some time to complete because it is running fio and
+/// then comparing the contents of the children to make sure they are in-sync.
+#[tokio::test]
+async fn rebuild_with_load() {
+    init_tracing();
+    let test = start_infrastructure("rebuild_with_load").await;
+    let nexus_hdl = &mut test.grpc_handle("ms1").await.unwrap();
+    let ms2 = &mut test.grpc_handle("ms2").await.unwrap();
+    let ms3 = &mut test.grpc_handle("ms3").await.unwrap();
+
+    const CHILD_SIZE_MB: u64 = 100;
+
+    // Create a nexus with 1 child.
+    let child1 =
+        bdev_create_and_share(ms2, Some(CHILD_SIZE_MB), Some("disk1".into()))
+            .await;
+    create_nexus(nexus_hdl, vec![child1.clone()]).await;
+
+    // Connect to nexus over NVMf.
+    let nexus_uri = publish_nexus(nexus_hdl).await;
+    let nexus_tgt = nvmf_connect(nexus_uri.clone());
+
+    // Run fio against nexus.
+    let (s, r) = unbounded::<i32>();
+    let nvmf_tgt = nexus_tgt.clone();
+    std::thread::spawn(move || {
+        s.send(common::fio_verify_size(&nvmf_tgt, NEXUS_SIZE))
+            .unwrap();
+    });
+
+    // Let fio run for a bit.
+    std::thread::sleep(Duration::from_secs(2));
+
+    // Add a child and rebuild.
+    let child2 =
+        bdev_create_and_share(ms3, Some(CHILD_SIZE_MB), Some("disk2".into()))
+            .await;
+    add_child(nexus_hdl, &child2, true).await;
+
+    // Wait for fio to complete
+    let fio_result = r.recv().unwrap();
+    assert_eq!(fio_result, 0, "Failed to run fio_verify_size");
+
+    // Wait for rebuild to complete.
+    assert!(
+        wait_for_rebuild_completion(nexus_hdl, &child2, Duration::from_secs(1))
+            .await
+    );
+
+    // Disconnect and destroy nexus
+    nvmf_disconnect(nexus_uri);
+    destroy_nexus(nexus_hdl).await;
+
+    // Check children are in-sync.
+    let child1_tgt = nvmf_connect(child1.clone());
+    let child2_tgt = nvmf_connect(child2.clone());
+    common::compare_devices(&child1_tgt, &child2_tgt, CHILD_SIZE_MB, true);
+    nvmf_disconnect(child1);
+    nvmf_disconnect(child2);
```

Comment:
@gila now that we're using nvmf more throughout the cargo tests, should NvmeTarget have a drop where it disconnects to prevent leaking devices on CI?

Reply:
ah but nvmeadm is also used by the csi so probably not a good idea then...